(chore) UI Documentation (#309)
kyma-project/cloud-manager: docs/contributor/ui/README.md

# Busola UI Documentation

> Kyma manages its modules' UI through the use of a ConfigMap. The ConfigMap has five parts. They are:

Reviewer: ```suggestion
Kyma manages its modules' user interfaces (UIs) using ConfigMaps. Each ConfigMap has five parts:
```

* `general` - used to describe the CRD the UI should be looking for as well as a description of the resource in the case none exists on the cluster.
* `list` - used to set which columns to display in the table view
* `detail` - used to display the detailed view of a single resource
* `form` - used to define a GUI form when creating a new resource or editing an existing one. 
* `translations` - used to define elements in different languages (English, German, etc.)

Examples and important notes of each part are below

### General
```yaml
resource:
  kind: GcpNfsVolume
  group: cloud-resources.kyma-project.io
  version: v1beta1
urlPath: gcpnfsvolumes
name: GCP NFS Volumes
scope: namespace
category: Storage
icon: shelf
description: >-
  GcpNfsVolume description here
```
It is imperative that `resource.kind` `resource.group` and `resource.version` matches its `CustomResourceDefinition`. If

> there are no matches, Busola will not render the UI and path rendering the resource inaccessible in Busola.

Reviewer: ```suggestion
there are no matches, Busola doesn't render the UI and path rendering the resource inaccessible in Busola.
```

### List
```yaml
- source: spec.fileShareName
  name: spec.fileShareName
  sort: true
- source: spec.location
  name: spec.location
  sort: true
- source: spec.tier
  name: spec.tier
  sort: true
- source: status.state
  name: status.state
  sort: true
```

> The `source` field is where busola will pull information from the monitored `CustomResource`.

Reviewer: ```suggestion
The `source` field is where Busola pulls information from the monitored custom resource (CR).
```

> The `name` field is the human-readable name for the field. This field will look up the value in `Translation` and replace it with its found value.

Reviewer: ```suggestion
The `name` field is the human-readable name for the field. This field looks up the value in `Translation` and replaces it with its found value.
```

> If no value is found in `Translation`, it will display as is.

Reviewer: ```suggestion
If no value is found in `Translation`, it displays as is.
```

> For example, if we have `name: spec.location`, it will go to [translations](#translations), lookup `spec.location` and replace it with `Location`.

Reviewer: ```suggestion
For example, if we have `name: spec.location`, it goes to [translations](#translations), looks up `spec.location` and replaces it with `Location`.
```

### Form
```yaml
- path: spec.capacityGb
  simple: true
  name: spec.capacityGb
  required: true
- path: spec.fileShareName
  simple: true
  name: spec.fileShareName
  required: true
- path: spec.location
  simple: true
  name: spec.location
  required: true
- path: spec.tier
  simple: true
  name: spec.tier
  required: true
```

`simple` is a boolean used to display the field in the simple fom. By default, it is `false`

[Official Forms Documentation](https://github.com/kyma-project/busola/blob/main/docs/extensibility/40-form-fields.md)

<a id="translations"></a>
### Translations
```yaml
en:
  spec.tier: Tier
  spec.location: Location
  spec.fileShareName: File Share Name
  spec.capacityGb: Capacity (Gb)
  spec.ipRange: IP Range
  configuration: Configuration
  status.state: State
  status: Status
```

`translations` is optional, but it supports languages formatted for [i18next](https://www.i18next.com/). Translations prettify the `name` field of the aformentioned sections.
They are key-value pairs.

[Official Translations Documentation](https://github.com/kyma-project/busola/blob/main/docs/extensibility/translations-section.md)

# Generating the ConfigMap

> The generation of ConfigMap is handled through `kustomize`. Kustomize  combines general, list, detail, form, and translations

Reviewer: ```suggestion
The generation of a ConfigMap is handled using `kustomize`. Kustomize  combines general, list, detail, form, and translations
```

into a single ConfigMap.

> Because each CustomResource will have its own UI that must be generated, there is a single macro in the `Makefile` to generate all the UI.

Reviewer: ```suggestion
Because each CustomResource has its own UI that must be generated, there is a single macro in the `Makefile` to generate all the UIs.
```

`make build_ui`

For every new UI component, a new `kustomize` command should be added to the Makefile `build_ui` macro.

[Official kustomize documentation](https://kustomize.io/)

# Deploying to your desired provider (AWS, GCP, Azure)

Currently, the pipeline looks at the presence of ConfigMaps files in specific directories to deploy. They are the following:

AWS - `config/dist/skr/crd/bases/providers/aws`

Azure - `config/dist/skr/crd/bases/providers/azure`

GCP - `config/dist/skr/crd/bases/providers/gcp`

Instead of manually copying over the ConfigMap into the specified directory, use the bash script `sync.sh`  

Like the `Makefile`, any new UI component should be added to `sync.sh`.

# Testing your UI changes without using the pipeline

Because the UI is dictated by ConfigMaps, you can deploy directly to a cluster with kyma installed. Just ensure your ConfigMap goes into the `kyma-system`
namespace. 

> You can then `kubectl apply` your ConfigMap

Reviewer: ```suggestion
You can then `kubectl apply` your ConfigMap.
```

> If a matching CustomResource is found, Busola will render the UI and path.

Reviewer: ```suggestion
If a matching CustomResource is found, Busola renders the UI and path.
```
